src/Datadog.Trace/HttpOverStreams/DatadogHttpClient.cs

```diff
@@ -0,0 +1,146 @@
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.HttpOverStreams
+{
+    internal class DatadogHttpClient
+    {
+        private const int BufferSize = 10240;
+        private static readonly Vendors.Serilog.ILogger Logger = DatadogLogging.For<DatadogHttpClient>();
+
+        public async Task<HttpResponse> SendAsync(HttpRequest request, Stream requestStream, Stream responseStream)
+        {
+            await SendRequest(request, requestStream).ConfigureAwait(false);
+            return await ReadResponse(responseStream).ConfigureAwait(false);
+        }
+
+        private async Task SendRequest(HttpRequest request, Stream requestStream)
+        {
+            // TODO: Determine if it's always ASCII
+            using (var writer = new StreamWriter(requestStream, Encoding.ASCII, BufferSize, leaveOpen: true))
+            {
+                await DatadogHttpHeaderHelper.WriteLeadingHeaders(request, writer).ConfigureAwait(false);
+
+                foreach (var header in request.Headers)
+                {
+                    await DatadogHttpHeaderHelper.WriteHeader(writer, header).ConfigureAwait(false);
+                }
+
+                await DatadogHttpHeaderHelper.WriteEndOfHeaders(writer).ConfigureAwait(false);
+            }
+
+            await request.Content.CopyToAsync(requestStream).ConfigureAwait(false);
+            Logger.Debug("Datadog HTTP: Flushing stream.");
+            await requestStream.FlushAsync().ConfigureAwait(false);
+        }
+
+        private Task<HttpResponse> ReadResponse(Stream responseStream)
+        {
+            var headers = new HttpHeaders();
+            int statusCode = 0;
+            string responseMessage = null;
+
+            // hack: buffer the entire response so we can seek
+            var memoryStream = new MemoryStream();
+            responseStream.CopyTo(memoryStream);
+            memoryStream.Position = 0;
+            int streamPosition = 0;
+
+            using (var reader = new StreamReader(memoryStream, Encoding.ASCII, detectEncodingFromByteOrderMarks: false, BufferSize, leaveOpen: true))
+            {
+                // https://tools.ietf.org/html/rfc2616#section-4.2
+                // HTTP/1.1 200 OK
+                // HTTP/1.1 XXX MESSAGE
+                string line = reader.ReadLine();
+                streamPosition += reader.CurrentEncoding.GetByteCount(line) + DatadogHttpHeaderHelper.CrLfLength;
+
+                const int statusCodeStart = 9;
+                const int statusCodeLength = 3;
+                const int startOfMessage = 13;
+
+                if (!int.TryParse(line.Substring(statusCodeStart, statusCodeLength), out statusCode))
+                {
+                    throw new DatadogHttpRequestException("Invalid response, can't parse status code. Line was:" + line);
+                }
+
+                responseMessage = line.Substring(startOfMessage);
+
+                // TODO: Get these from a StringBuilderCache
```

Comment:
👍

Reply:
I am going to do a follow up PR so pull the StringBuilderCache into a shared location and update this.